src/cdk/tree/tree.ts

```diff
@@ -311,6 +571,513 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
+   */
+  private _findParentForNode(node: T, index: number, cachedNodes: readonly T[]): T | null {
+    // In all cases, we have a mapping from node to level; all we need to do here is backtrack in
+    // our flattened list of nodes to determine the first node that's of a level lower than the
+    // provided node.
+    if (!cachedNodes.length) {
+      return null;
+    }
+    const currentLevel = this._levels.get(this._getExpansionKey(node)) ?? 0;
+    for (let parentIndex = index - 1; parentIndex >= 0; parentIndex--) {
+      const parentNode = cachedNodes[parentIndex];
+      const parentLevel = this._levels.get(this._getExpansionKey(parentNode)) ?? 0;
+
+      if (parentLevel < currentLevel) {
+        return parentNode;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Given a set of root nodes and the current node level, flattens any nested
+   * nodes into a single array.
+   *
+   * If any nodes are not expanded, then their children will not be added into the array.
+   * NB: this will still traverse all nested children in order to build up our
+   * internal data models, but will not include them in the returned array.
+   */
+  private _flattenNestedNodesWithExpansion(nodes: readonly T[], level = 0): Observable<T[]> {
+    const childrenAccessor = this._getChildrenAccessor();
+    // If we're using a level accessor, we don't need to flatten anything.
+    if (!childrenAccessor) {
+      return observableOf([...nodes]);
+    }
+
+    return observableOf(...nodes).pipe(
+      concatMap(node => {
+        const parentKey = this._getExpansionKey(node);
+        if (!this._parents.has(parentKey)) {
+          this._parents.set(parentKey, null);
+        }
+        this._levels.set(parentKey, level);
+
+        const children = coerceObservable(childrenAccessor(node));
+        return concat(
+          observableOf([node]),
+          children.pipe(
+            take(1),
+            tap(childNodes => {
+              this._ariaSets.set(parentKey, [...(childNodes ?? [])]);
+              for (const child of childNodes ?? []) {
+                const childKey = this._getExpansionKey(child);
+                this._parents.set(childKey, node);
+                this._levels.set(childKey, level + 1);
+              }
+            }),
+            switchMap(childNodes => {
+              if (!childNodes) {
+                return observableOf([]);
+              }
+              return this._flattenNestedNodesWithExpansion(childNodes, level + 1).pipe(
+                map(nestedNodes => (this.isExpanded(node) ? nestedNodes : [])),
+              );
+            }),
+          ),
+        );
+      }),
+      reduce((results, children) => {
+        results.push(...children);
+        return results;
+      }, [] as T[]),
+    );
+  }
+
+  /**
+   * Converts children for certain tree configurations.
+   *
+   * This also computes parent, level, and group data.
+   */
+  private _convertData(
```

Comment:
This function name (and description) is pretty vague- could it be more specific?

Reply:
That this method actual does is map data from the DataSource into `RenderingData` type.

It also needs to handle all the different cases for nodeType. The Tree infers the type of the nodes based on the data its DataSource gives it. Handling nodeType is part of why there are so many conditional cases here.